mlkem/poly.c

```diff
@@ -82,14 +81,43 @@ __CPROVER_ensures(__CPROVER_return_value < 32)
  *                 to be decompressed.
  ************************************************************/
 uint32_t scalar_decompress_q_32(uint32_t u)
-/* INDENT-OFF */
-__CPROVER_requires(0 <= u && u < 32)
-__CPROVER_ensures(__CPROVER_return_value < KYBER_Q)
-/* INDENT-ON */
 {
     return ((u * KYBER_Q) + 16) / 32;
 }
 
+/************************************************************
+ * Name: coeff_signed_to_unsigned
+ *
+ * Description: converts signed polynomial coefficient
+ *              from signed (-3328 .. 3328) form to
+ *              unsigned form (0 .. 3328).
+ *
+ * Note: Cryptographic constant time implementation
+ *
+ * Examples:       0 -> 0
+ *                 1 -> 1
+ *              3328 -> 3328
+ *                -1 -> 3328
+ *                -2 -> 3327
+ *             -3328 -> 1
+ *
+ * Arguments: c: signed coefficient to be converted
+ ************************************************************/
+uint16_t coeff_signed_to_unsigned (int16_t c)
+{
+    int32_t r = (int32_t) c;
+
+    // Add Q if r is negative
+    int32_t factor = (r < 0); // 1 if r < 0; 0 if r >= 0
+    r = r + (factor * KYBER_Q);
```

Comment:
I think it's not unlikely that a compiler would introduce a branch here. One way to thwart this is by calculating `r` and `r + KYBER_q` and having a `safe_cond_select()` stub which, based on either (a) a mask that's 0 or -1, or (b) an int that's 0 or 1, picks one of the two arguments. This function would itself be a C stub that's ideally replaced by ASM.

Reply:
Agreed - we need a translation unit of safe constant-time functions like cond_select() and others. Disagree about the third parameter - it's a Boolean not a bleedin' Integer!

Reply:
The multiplication `factor*KYBER_Q` you want will disappear when compiler optimize it. 
https://godbolt.org/z/xaY39sdWr 
So the compiler see our intentions and convert to:
`r += (c >> 15) & KYBER_Q` or `r += (c >> 31) & KYBER_Q`.
On other platform like AVR it's a branch condition. 
I think the new style and the old style are equivalent in the eye of compiler.

Reply:
The semantics of the >> operator on a signed, negative integer is implementation-defined (see WG14 N2310 6.5.7(5) for example), so I prefer not to use it. Strictly speaking, the use of >> here defies verification with CBMC.